components/raftstore/src/store/util.rs

```diff
@@ -1123,77 +1129,98 @@ pub fn check_conf_change(
     }
 }
 
-fn check_remove_or_demote_voter(
+/// Check the would-be availability if the operation proceed.
+/// If the slow peers count would be equal or larger than normal peers count,
+/// then the operations would be rejected
+fn check_availability_by_last_heartbeats(
     region: &metapb::Region,
     cfg: &Config,
     change_peers: &[ChangePeerRequest],
     leader_id: u64,
-    peer_heartbeat: &collections::HashMap<u64, std::time::Instant>,
+    peer_heartbeats: &collections::HashMap<u64, std::time::Instant>,
 ) -> Result<()> {
-    let mut slow_voters_count = 0;
-    let mut normal_voters_count = 0;
+    let mut slow_voters = vec![];
+    let mut normal_voters = vec![];
+
     // Here we assume if the last beartbeat is within 2 election timeout, the peer
     // is healthy. When a region is hibernate, we expect all its peers are *slow*
     // and it would still allow the operation
     let slow_voter_threshold =
         2 * cfg.raft_base_tick_interval.0 * cfg.raft_max_election_timeout_ticks as u32;
-    for (id, last_heartbeat) in peer_heartbeat {
+    for (id, last_heartbeat) in peer_heartbeats {
         // for slow and normal peer calculation, we only count voter role
         if region
             .get_peers()
             .iter()
             .find(|p| p.get_id() == *id)
-            .map_or(false, |p| p.role == PeerRole::Voter)
+            .map_or(false, |p| {
+                p.role == PeerRole::Voter || p.role == PeerRole::IncomingVoter
+            })
         {
             // leader itself is not a slow peer
             if *id == leader_id || last_heartbeat.elapsed() <= slow_voter_threshold {
-                normal_voters_count += 1;
+                normal_voters.push(*id);
             } else {
-                slow_voters_count += 1;
+                slow_voters.push(*id);
             }
         }
     }
 
+    let is_healthy = normal_voters.len() > slow_voters.len();
```

Comment:
One more tip, if `is_healthy` already is `false`, the following steps for checking could be skipped.